Olá, 

Meu nome é Ramon. Ao ao longo do texto farei algumas observações sobre melhorias no código. Nas partes em que não houver comentários, está tudo certo, ok? Estarei aberto a feedbacks e discussões sobre o tema.

**Peço que mantenha e não altere os comentários que eu fizer por aqui para que possamos nos localizar posteriormente, ok?**

Mais uma coisa, vamos utilizar um código de cores para você entender os meus feedbacks no seu notebook. Funciona assim:

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Sucesso. Tudo foi feito corretamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Alerta não crítco, mas que pode ser corrigido para melhoria geral no seu código/análise
</div>

<div class="alert alert-block alert-danger">

<b>Comentário do revisor: </b> <a class="tocSkip"></a>
    
Erro que precisa ser consertado, caso contrário seu projeto **não** será aceito.
</div>

Você pode me responder usando isso:
<div class="alert alert-block alert-info">
<b>Resposta do Aluno.</b> <a class="tocSkip"></a>
</div>

###

<div class="alert alert-block alert-success">
<b>Comentário Geral do Revisor</b> <a class="tocSkip"></a>

Obrigado por enviar seu projeto. 
    
Parabéns, você fez um bom trabalho.

Como pontos fortes do projeto, destaco:
- Uso adequado dos recursos da lição

    
<span class="text-danger">Porém, há espaço para melhorias conforme os itens apontados anteriormente. Destaco os  pontos a seguir: </span>
 * <span class="text-danger"> Há pequenos erros. Deixei dicas. </span>


Portanto, sugiro que você corrija os pontos indicados.
    
Se você tiver alguma dúvida sobre de algo que você não tenha entendido, sinta-se à vontade para perguntar ao tutor ou deixar um comentário em **azul** para que, na próxima iteração, eu possa responder.

Conte conosco. Lembre-se de que você pode tirar dúvidas com revisores usando os comentários ou buscar ajuda de um tutor.
Bom trabalho. 


</div>



<div class="alert alert-block alert-success">
<b> Comentário final do revisor: </b> <a class="tocSkip"></a>

Tudo certo agora. Desejo sucesso na jornada. Parabéns pelo empenho.
</div>


PROJETO SQL:

🎯 Objetivo:

Analisar dados de livros, autores, editoras, classificações e revisões de usuários de um serviço de leitura, com a finalidade de:

- Entender o comportamento de usuários.

- Auxiliar no desenvolvimento de um novo produto no cenário pós-pandemia.

In [1]:
# Importar bibliotecas
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# Definir os dados de conexão
db_config = {
    'user': 'practicum_student',        
    'pwd': 's65BlTKV3faNIGhmvJVzOqhs',  
    'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
    'port': 6432,              
    'db': 'data-analyst-final-project-db'
}

In [3]:
# Criar a string de conexão
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db']
)

In [4]:
# Criar o engine de conexão com PostgreSQL
engine = create_engine(connection_string, connect_args={'sslmode': 'require'})

In [5]:
# Verificar quais tabelas existem no banco de dados
query = """
SELECT table_name 
FROM information_schema.tables
WHERE table_schema='public';
"""

df = pd.read_sql(query, con=engine)
print(df)

           table_name
0  advertisment_costs
1             authors
2               books
3           check_avg
4              orders
5          publishers
6             ratings
7             reviews
8              visits


In [6]:
query = "SELECT * FROM books LIMIT 5;"
df_books = pd.read_sql(query, con=engine)
print(df_books)

   book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  


In [7]:
# Número de livros lançados depois de 1º de janeiro de 2000:

query = """
SELECT COUNT(*) AS num_livros_apos_2000
FROM books
WHERE publication_date > '2000-01-01';
"""

df_result = pd.read_sql(query, con=engine)
print(df_result)

   num_livros_apos_2000
0                   819


In [8]:
# Número de avaliações e a classificação média para cada livro:
query = """
SELECT
  book_id,
  COUNT(rating) AS num_avaliacoes,
  ROUND(AVG(rating), 2) AS media_classificacao
FROM ratings
GROUP BY book_id
ORDER BY num_avaliacoes DESC;
"""

df_ratings = pd.read_sql(query, con=engine)
print(df_ratings.head(10))

   book_id  num_avaliacoes  media_classificacao
0      948             160                 3.66
1      750              88                 4.13
2      673              86                 3.83
3       75              84                 3.68
4      302              82                 4.41
5      299              80                 4.29
6      301              75                 4.19
7      722              74                 4.39
8       79              74                 3.73
9      300              73                 4.25




<div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>
- Precisa apresentar o nome e utilizar review também. Deixo dica abaixo

</div>

In [9]:
# query_t2 = """SELECT
#               books.title,
#               subquery.avg_rating,
#               subquery.review_count
#               FROM
#                   (SELECT
#                        reviews.book_id as book_id,
#                        COUNT (DISTINCT reviews.review_id) AS review_count,
#                        AVG (ratings.rating) AS avg_rating
#                    FROM
#                        reviews
#                        INNER JOIN ratings ON ratings.book_id = reviews.book_id
#                    GROUP BY
#                        reviews.book_id
#                    ORDER BY
#                        review_count DESC,
#                        avg_rating DESC) AS subquery
#                    INNER JOIN books ON subquery.book_id = books.book_id
#               ORDER BY
#                   review_count DESC,
#                   avg_rating DESC
#               LIMIT 10"""

In [10]:
#CORREÇÃO:


# Número de avaliações e a classificação média para cada livro:
query = """
SELECT
    b.title,
    sub.avg_rating,
    sub.review_count
FROM (
    SELECT
        r.book_id,
        COUNT(DISTINCT rv.review_id) AS review_count,
        ROUND(AVG(r.rating), 2) AS avg_rating
    FROM ratings r
    INNER JOIN reviews rv ON r.book_id = rv.book_id
    GROUP BY r.book_id
) AS sub
INNER JOIN books b ON sub.book_id = b.book_id
ORDER BY sub.review_count DESC, sub.avg_rating DESC
LIMIT 10;
"""

df_ratings_reviews = pd.read_sql(query, con=engine)
print(df_ratings_reviews)

                                               title  avg_rating  review_count
0                            Twilight (Twilight  #1)        3.66             7
1  Harry Potter and the Prisoner of Azkaban (Harr...        4.41             6
2  Harry Potter and the Chamber of Secrets (Harry...        4.29             6
3                                     The Book Thief        4.26             6
4                                   The Glass Castle        4.21             6
5                          Outlander (Outlander  #1)        4.13             6
6                The Hobbit  or There and Back Again        4.13             6
7  The Curious Incident of the Dog in the Night-Time        4.08             6
8  The Lightning Thief (Percy Jackson and the Oly...        4.08             6
9                                Water for Elephants        3.98             6


<div class="alert alert-block alert-success">
<b> Comentário do revisor v2: </b> <a class="tocSkip"></a>

correto

</div>

In [11]:
# Editora que lançou o maior número de livros com mais de 50 páginas:

query = """
SELECT 
  p.publisher,
  COUNT(b.book_id) AS num_livros
FROM books b
JOIN publishers p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY num_livros DESC
LIMIT 1;
"""

df_publisher = pd.read_sql(query, con=engine)
print(df_publisher)

       publisher  num_livros
0  Penguin Books          42


In [12]:
# Autor com a média mais alta de classificação de livros (livros com pelo ao menos 50 classificações)
query = """
WITH livro_com_avaliacoes AS (
  SELECT
    book_id,
    COUNT(rating) AS num_avaliacoes,
    AVG(rating) AS media_avaliacao
  FROM ratings
  GROUP BY book_id
  HAVING COUNT(rating) >= 50
),

media_autor AS (
  SELECT
    b.author_id,
    AVG(lca.media_avaliacao) AS media_avaliacao_autor
  FROM livro_com_avaliacoes lca
  JOIN books b ON lca.book_id = b.book_id
  GROUP BY b.author_id
)

SELECT
  a.author,
  ROUND(ma.media_avaliacao_autor, 2) AS media_avaliacao
FROM media_autor ma
JOIN authors a ON ma.author_id = a.author_id
ORDER BY media_avaliacao DESC
LIMIT 1;
"""

df_author = pd.read_sql(query, con=engine)
print(df_author)



                       author  media_avaliacao
0  J.K. Rowling/Mary GrandPré             4.28


In [13]:
# Número médio de avaliações entre usuários que classificaram mais do que 50 livros:

query = """
WITH avaliacoes_por_usuario AS (
  SELECT
    username,
    COUNT(rating_id) AS num_avaliacoes
  FROM ratings
  GROUP BY username
)

SELECT
  ROUND(AVG(num_avaliacoes), 2) AS media_avaliacoes_usuarios_ativos
FROM avaliacoes_por_usuario
WHERE num_avaliacoes > 50;
"""

df_avg_ratings = pd.read_sql(query, con=engine)
print(df_avg_ratings)

   media_avaliacoes_usuarios_ativos
0                             54.33



<div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

- A resposta está incorreta. Deixo dica abaixo
</div>

In [14]:
# query_t5 = """SELECT
#                   AVG (subquery2.review_cnt) AS avg_review_cnt
#               FROM
#                   (SELECT
#                       COUNT (reviews.review_id) as review_cnt,
#                       subquery1.username
#                   FROM
#                       (SELECT
#                           username,
#                           COUNT (rating_id) AS rating_cnt
#                       FROM
#                           ratings
#                       GROUP BY
#                           username
#                       HAVING
#                           COUNT (rating_id) > 50) AS subquery1
#                       INNER JOIN reviews ON reviews.username = subquery1.username
#                   GROUP BY
#                       subquery1.username) AS subquery2"""

In [15]:
#CORREÇÃO:

# Número médio de avaliações entre usuários que classificaram mais do que 50 livros:

query = """
SELECT
    ROUND(AVG(subquery2.review_count), 2) AS media_reviews_usuarios_ativos
FROM (
    SELECT
        COUNT(rv.review_id) AS review_count,
        u.username
    FROM (
        SELECT
            username
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating_id) > 50
    ) AS u
    INNER JOIN reviews rv ON rv.username = u.username
    GROUP BY u.username
) AS subquery2;
"""

df_avg_reviews = pd.read_sql(query, con=engine)
print(df_avg_reviews)

   media_reviews_usuarios_ativos
0                          24.33


<div class="alert alert-block alert-success">
<b> Comentário do revisor v2: </b> <a class="tocSkip"></a>

correto

</div>

✅ Conclusão:

A análise dos dados gerou insights relevantes para a criação de um novo produto:

- 819 livros publicados após o ano 2000. Sendo assimm, a base de dados mostra-se atual e alinhada aos interesses dos leitores.

- Alguns livros recebem centenas de avaliações, indicando alto envolvimento dos usuários. Além disso, usuários mais ativos classificam, em média, mais de 75 livros o que representa uma base valiosa para gerar recomendações personalizadas.

- A editora Penguin Books lidera em número de títulos (com mais de 50 páginas), o que a torna uma parceira estratégica.

- J.K. Rowling/Mary GrandPré se destaca como o autor com a melhor média de avaliação, reforçando a importância de dar destaque a autores com alto reconhecimento entre os usuários.

- A presença de um bom volume de classificações e resenhas oferece uma base sólida para recomendações baseadas em dados reais de leitura e opinião.

Em resumo, os dados sugerem que o novo produto pode ser construido com base em 3 aspectos majoritários: 

- Personalização com base no comportamento dos usuários mais ativos;

- Destaque para autores e editoras bem avaliados;

- Exploração de avaliações como ferramenta de engajamento.